File: morphir-dotnet/src/Morphir.CodeModel/IR/Documented.fs
module Morphir.IR.Documented

/// <summary>
/// Type that represents a value that is documented value.
/// </summary>
type Documented<'T> = { Doc: string; Value: 'T }

/// <summary>
/// Maps a function over the value of a documented value.
/// </summary>
[<CompiledName("Map")>]
let map (f: 'a -> 'b) (d: Documented<'a>) = { Doc = d.Doc; Value = f d.Value }

[<CompiledName("Value")>]
let inline value (d: Documented<'a>) = d.Value

let inline documented (doc: string) (value: 'a) : Documented<'a> = { Doc = doc; Value = value }

let (|Doc|) documented = documented.Doc
let (|Value|) documented = documented.Value

File: morphir/ir/core/src/morphir/ir/Documented.scala
package morphir.ir

/** Generated based on IR.Documented
*/
object Documented{

  final case class Documented[A](
    doc: morphir.sdk.String.String,
    value: A
  ){}
  
  def map[A, B](
    f: A => B
  )(
    d: morphir.ir.Documented.Documented[A]
  ): morphir.ir.Documented.Documented[B] =
    (morphir.ir.Documented.Documented(
      d.doc,
      f(d.value)
    ) : morphir.ir.Documented.Documented[B])

}